bndtools.builder/src/org/bndtools/builder/BndtoolsBuilder.java

```diff
@@ -148,13 +148,24 @@
 						if (!Files.isSameFile(javaPath.getParent(), wsPath)) {
 							msg = "The project directory: " + javaPath + " is not a subdirectory of the bnd workspace "
 								+ wsPath;
+						} else {
+							Path bndPath = javaPath.resolve(Project.BNDFILE);
+
+							if (!Files.exists(bndPath)) {
+								msg = "The project does not have a " + Project.BNDFILE + " file";
+							} else if (!Files.isReadable(bndPath)) {
+								msg = Project.BNDFILE + " is not readable";
+							} else if (Files.isDirectory(bndPath)) {
+								msg = Project.BNDFILE + " is a directory";
+							} else if (!Files.isRegularFile(bndPath)) {
+								msg = Project.BNDFILE + " is not a regular file";
```

Comment:
This is unreachable. If the file exists and is not a directory, then it must be a regular file. There are no other choices since you follow symlinks.

Reply:
It could be a pipe or a device or some other special file?

Reply:
How could the `bnd.bnd` path be a special file? I don't see how that could ever be the case. I think you want to test for the normal case first (isRegularFile) and it that fails, then try to make a nice error message.

Reply:
> How could the `bnd.bnd` path be a special file? I don't see how that could ever be the case. I think you want to test for the normal case first (isRegularFile) and it that fails, then try to make a nice error message.

It could be a special file if the user does something a bit silly or tries to be a bit smart. Admittedly, this is not an expected scenario. But I think it's a good rule of UI design to be able to handle the user doing something silly and/or unexpected and to give them the most useful feedback possible to help them fix the problem, especially if it is relatively simple and inexpensive to do so. In this case, the cost of the extra test is low (both in execution overhead and code size) and it's not in a performance-critical part of the code, so I'm struggling to see the downside in including it. But if you insist, I will remove the specific check for a directory.

There is already a "nice" error message (or at least, as nice as it can be) as the default:
https://github.com/bndtools/bnd/blob/0a3846f7f550ad0e2014c6f444bd760e5105993b/bndtools.builder/src/org/bndtools/builder/BndtoolsBuilder.java#L133

If none of the other diagnostic checks finds any more specific cause, this is what will be shown. If you can think of a better generic message, or think of more diagnostic tests that could be done, then I'm all ears!